cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,264 @@
+
+
+@value.value_equality(distinct_child_types=True)
+class GateFamily:
+    """Wrapper around gate instances/types describing a set of accepted gates.
+
+    By default, GateFamily supports initialization and containment for
+        a) Non-parameterized `cirq.Gate` instances.
+        b) Gate types deriving from `cirq.Gate`.
+
+    In order to create gate families with constraints on parameters of a gate
+    type, users should derive from the GateFamily class and override the
+    predicate method used to check for containment.
+    """
+
+    def __init__(
+        self,
+        gate: Union[Type[raw_types.Gate], raw_types.Gate],
+        *,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> None:
+        """Init GateFamily
+        Args:
+            gate: A gate type or a non-parameterized instance of a gate type.
+            name: The name of the gate family.
+            description: Human readable description of the gate family.
+
+        Raises:
+            ValueError: if `gate` is not a `cirq.Gate` instance or subclass.
+        """
+        if not (
+            isinstance(gate, raw_types.Gate)
+            or (isinstance(gate, type) and issubclass(gate, raw_types.Gate))
+        ):
+            raise ValueError(f'Gate {gate} must be an instance or subclass of `cirq.Gate`.')
+        if isinstance(gate, raw_types.Gate) and protocols.is_parameterized(gate):
+            raise ValueError(f'Gate {gate} must be a non-parameterized instance of `cirq.Gate`.')
+
+        self._gate = gate
+        self._name = name if name else self._default_name()
+        self._description = description if description else self._default_description()
+
+    def _gate_str(self, gettr: Callable[[Any], str] = str) -> str:
+        return (
+            gettr(self.gate)
+            if isinstance(self.gate, raw_types.Gate)
+            else f'{self.gate.__module__}.{self.gate.__name__}'
+        )
+
+    def _default_name(self) -> str:
+        if isinstance(self.gate, raw_types.Gate):
+            return f'Instance GateFamily: {self._gate_str()}'
+        else:
+            return f'Type GateFamily: {self._gate_str()}'
+
+    def _default_description(self) -> str:
+        if isinstance(self.gate, raw_types.Gate):
+            return f'Accepts `cirq.Gate` instances `g` s.t. `g == {self._gate_str()}`.'
+        else:
+            return f'Accepts `cirq.Gate` instances `g` s.t. `isinstance(g, {self._gate_str()})`.'
+
+    @property
+    def gate(self) -> Union[Type[raw_types.Gate], raw_types.Gate]:
+        return self._gate
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def description(self) -> str:
+        return self._description
+
+    def predicate(self, g: raw_types.Gate) -> bool:
+        """Checks whether gate instance `g` belongs to this GateFamily."""
+        return g == self.gate if isinstance(self.gate, raw_types.Gate) else isinstance(g, self.gate)
```

Comment:
Well, The type of `self.gate` is `Union[Type[raw_types.Gate], raw_types.Gate]`. So if the type of `self.gate` is 

a) `raw_types.Gate`, the method will check for gate comparison (if condition will trigger). 
b) `Type[raw_types.Gate]`, the method will do an `isinstance` check, but it'll pass only if `g` is an instance of `raw_types.Gate` because type of `self.gate` is `Type[raw_types.Gate]`. Hence, the method will not accept arbitrary types. 


The reason I've chosen to support only gate containment check here is to make it harder for users writing new derived gate families to make mistakes. Supporting operations in this method would require that users are aware of the complex operation and gate relationships, which is not always true. I want to avoid users writing checks like `isinstance(op, GateOperation)` for validation, which is currently present all over the library. Ofc, users are free to override the `__contains__` method if they need to handle certain operations in a special way.